File: Service/Translators/ClientGiftTranslator.vb
Imports DAL
Imports BO
Imports System.Text.RegularExpressions
Public Class ClientGiftTranslator

    Friend Shared Function TranslateEntityToBO(_entity As ClientGift, bo As ClientGiftBO) As ErrorCode
        If _entity Is Nothing Then Return ErrorCode.EntityNull
        If bo Is Nothing Then Return ErrorCode.BoNull
        bo.Id = _entity.Id
        bo.AccountId = _entity.ClientAccountId
        bo.Description = _entity.Description
        If _entity.Activity IsNot Nothing Then
            For Each item In _entity.Activity
                Dim Activity As New ActivityBO
                Dim err = ActivityTranslator.TranslateEntityToBO(item, Activity)
                If err <> ErrorCode.Success Then Return err
                bo.Activities.Add(Activity)
            Next
        End If
        Return ErrorCode.Success
    End Function
    Friend Shared Function TranslateBOToEntity(_entity As ClientGift, bo As ClientGiftBO, uow As UnitOfWork) As ErrorCode
        If _entity Is Nothing Then Return ErrorCode.EntityNull
        If bo Is Nothing Then Return ErrorCode.BoNull
        _entity.ClientAccountId = bo.AccountId
        _entity.Description = bo.Description

        Dim err = HandleActivities(_entity, bo.Activities, uow)
        If (err <> ErrorCode.Success) Then Return err
        Return ErrorCode.Success
    End Function

    Private Shared Function HandleActivities(_entity As ClientGift, activities As List(Of ActivityBO), uow As UnitOfWork) As ErrorCode
        If (activities.Count = 0) Then Return ErrorCode.Success
        For Each x In activities
            If (x.ID = 0) Then
                'should never happen
            Else
                'add the activity to the clientgift
                If (Not _entity.Activity.Any(Function(m) m.ActivityID = x.ID)) Then
                    Dim act = uow.GetActivityDAO().GetById(x.ID)
                    _entity.Activity.Add(act)
                End If
            End If
        Next
        'delete
        Dim delList As New List(Of Activity)
        For Each x In _entity.Activity
            If (Not activities.Any(Function(f) f.ID = x.ActivityID)) Then
                delList.Add(x)
            End If
        Next
        For Each x In delList
            _entity.Activity.Remove(x)
        Next
        Return ErrorCode.Success
    End Function


End Class

File: ServiceCore/Translators/ClientGiftTranslator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using BOCore;
using DALCore;
using DALCore.Models;

namespace ServiceCore.Translators
{
    public class ClientGiftTranslator
    {
        internal static ErrorCode TranslateEntityToBO(ClientGift _entity, ClientGiftBO bo)
        {
            if (_entity == null)
                return ErrorCode.EntityNull;
            if (bo == null)
                return ErrorCode.BoNull;
            bo.Id = _entity.Id;
            bo.AccountId = _entity.ClientAccountId;
            bo.Description = _entity.Description;
            if (_entity.Activity != null)
            {
                foreach (var item in _entity.Activity)
                {
                    ActivityBO Activity = new ActivityBO();
                    var err = ActivityTranslator.TranslateEntityToBO(item, Activity);
                    if (err != ErrorCode.Success)
                        return err;
                    bo.Activities.Add(Activity);
                }
            }
            return ErrorCode.Success;
        }
        internal static ErrorCode TranslateBOToEntity(ClientGift _entity, ClientGiftBO bo, UnitOfWorkCore uow)
        {
            if (_entity == null)
                return ErrorCode.EntityNull;
            if (bo == null)
                return ErrorCode.BoNull;
            _entity.ClientAccountId = bo.AccountId;
            _entity.Description = bo.Description;

            var err = HandleActivities(_entity, bo.Activities, uow);
            if ((err != ErrorCode.Success))
                return err;
            return ErrorCode.Success;
        }

        private static ErrorCode HandleActivities(ClientGift _entity, List<ActivityBO> activities, UnitOfWorkCore uow)
        {
            if (_entity == null) return ErrorCode.EntityNull;
            if (uow == null) return ErrorCode.UowNull;

            // Null-safe
            activities ??= new List<ActivityBO>();
            _entity.Activity ??= new List<Activity>();

            // Gewenste IDs uit de BO (alleen geldige >0)
            var desiredIds = new HashSet<int>(activities.Where(a => a.ID > 0)
                                                        .Select(a => a.ID));

            // Bestaande IDs op de entity
            var existingIds = _entity.Activity.Select(a => a.ActivityId).ToHashSet();

            // ADD: wat nog niet aanwezig is
            foreach (var id in desiredIds)
            {
                if (!existingIds.Contains(id))
                {
                    var act = uow.Activities.GetById(id);
                    if (act != null)
                        _entity.Activity.Add(act);
                }
            }

            // DELETE: wat niet meer gewenst is
            var delList = _entity.Activity
                                 .Where(a => !desiredIds.Contains(a.ActivityId))
                                 .ToList();

            foreach (var a in delList)
                _entity.Activity.Remove(a);

            return ErrorCode.Success;
        }
    }
}
